Manually invalidate the FontCache before clearing ThreadGlobalData.

mattwoodrow opened the pull request:
#### 2d8a9204f770ff652583b404b5b3cf4a047cad17
<pre>
Manually invalidate the FontCache before clearing ThreadGlobalData.
<a href="https://bugs.webkit.org/show_bug.cgi?id=248502">https://bugs.webkit.org/show_bug.cgi?id=248502</a>

Reviewed by Cameron McCormack.

Destructing the FontCache can recurse back into the ThreadGlobalData getter (to remove cached entries), so
we want to manually clear the FontCache before clearing the ThreadGlobalData pointer.

This also moves m_destroyed to earlier in the class, so it&apos;s guaranteed to be destructed after m_fontCache.

* Source/WebCore/platform/ThreadGlobalData.cpp:
(WebCore::ThreadGlobalData::destroy):
* Source/WebCore/platform/ThreadGlobalData.h:
* Source/WebCore/platform/graphics/FontCache.h:

Canonical link: <a href="https://commits.webkit.org/257160@main">https://commits.webkit.org/257160@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5ca3e7b1156639091ccd8e558e5d616edb694c38

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/98082 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/7296 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/31232 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/107543 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/167809 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/102022 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/7784 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/36059 "Built successfully") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/36/builds/90616 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/104153 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/103734 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/5855 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/84679 "Built successfully") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/90616 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/87724 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/89438 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/36/builds/90616 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/1270 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/20871 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/82/builds/1234 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/22374 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4928 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/6095 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/44825 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/2535 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/41779 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

heycam commented on Source/WebCore/platform/ThreadGlobalData.h:
```diff
@@ -106,6 +106,8 @@ class ThreadGlobalData : public PAL::ThreadGlobalData {
     FontCache* fontCacheIfNotDestroyed() { return m_destroyed ? nullptr : &fontCache(); }
 
 private:
+    bool m_destroyed { false };
+
```
Nit: put this just before `m_cachedResourceRequestInitiators`?

mattwoodrow replied:
It's intentionally at the top, so that destruction order (last to first) guarantees that it's destroyed after the cache objects.

heycam replied:
I just mean in the same position relative to the other fields, but below the function declarations, i.e. just below `initializeFontCache()`.

mattwoodrow committed:
Manually invalidate the FontCache before clearing ThreadGlobalData.
https://bugs.webkit.org/show_bug.cgi?id=248502

Reviewed by Cameron McCormack.

Destructing the FontCache can recurse back into the ThreadGlobalData getter (to remove cached entries), so
we want to manually clear the FontCache before clearing the ThreadGlobalData pointer.

This also moves m_destroyed to earlier in the class, so it's guaranteed to be destructed after m_fontCache.

* Source/WebCore/platform/ThreadGlobalData.cpp:
(WebCore::ThreadGlobalData::destroy):
* Source/WebCore/platform/ThreadGlobalData.h:
* Source/WebCore/platform/graphics/FontCache.h:

Canonical link: https://commits.webkit.org/257160@main
